common/functions/src/scalars/conditionals/if.rs

```diff
@@ -11,59 +11,138 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+use std::sync::Arc;
 
-use std::fmt;
-
-use common_datavalues::columns::DataColumn;
-use common_datavalues::prelude::DataColumnsWithField;
-use common_datavalues::DataTypeAndNullable;
+use common_datavalues2::prelude::*;
+use common_datavalues2::remove_nullable;
+use common_datavalues2::type_coercion::compare_coercion;
 use common_exception::Result;
 
-use crate::scalars::function_factory::FunctionDescription;
+use crate::scalars::cast_column_field;
 use crate::scalars::function_factory::FunctionFeatures;
-use crate::scalars::Function;
+use crate::scalars::Function2;
+use crate::scalars::Function2Description;
+use crate::with_match_scalar_type;
 
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 pub struct IfFunction {
-    _display_name: String,
+    display_name: String,
 }
 
 impl IfFunction {
-    pub fn try_create_func(display_name: &str) -> Result<Box<dyn Function>> {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function2>> {
         Ok(Box::new(IfFunction {
-            _display_name: display_name.to_string(),
+            display_name: display_name.to_string(),
         }))
     }
 
-    pub fn desc() -> FunctionDescription {
-        FunctionDescription::creator(Box::new(Self::try_create_func))
-            .features(FunctionFeatures::default().deterministic().num_arguments(3))
+    pub fn desc() -> Function2Description {
+        let mut features = FunctionFeatures::default().num_arguments(3);
+        features = features.deterministic();
+        Function2Description::creator(Box::new(Self::try_create)).features(features)
     }
 }
 
-impl Function for IfFunction {
+impl Function2 for IfFunction {
     fn name(&self) -> &str {
         "IfFunction"
     }
 
-    fn return_type(&self, args: &[DataTypeAndNullable]) -> Result<DataTypeAndNullable> {
-        // IF(condition, value_if_true, value_if_false) is nullable if 'value_if_true' is nullable or 'value_if_false' is nullable.
-        // The condition of 'Null' is treated as false. SELECT IF(NULL, "YES", "NO") -> "No"
-        let nullable = args[1].is_nullable() || args[2].is_nullable();
+    fn return_type(&self, args: &[&DataTypePtr]) -> Result<DataTypePtr> {
+        let l_dt = remove_nullable(args[1]);
+        let r_dt = remove_nullable(args[2]);
 
-        let mut aggregate_args = Vec::with_capacity(args.len() - 1);
-        for arg in args.iter().skip(1) {
-            aggregate_args.push(arg.data_type().clone())
+        let mut least_supertype = compare_coercion(&l_dt, &r_dt)?;
```

Comment:
Here need use `aggregate_types` instead of `compare_coercion`

Reply:
thanks for reminding,, `aggregate_types` is better, I will make another pr to make it.